File: python/vista_api_client/api_resources/api_resource.py

from enum import Enum
from urllib.parse import urljoin

import requests


class HttpMethods(Enum):
    GET = 'GET'
    POST = 'POST'
    DELETE = 'DELETE'


class ApiResource():
    def __init__(self, secret, branch, hostname):
        self.hostname = hostname
        self.branch = branch
        self.secret = secret

    def dispatch(self, route, method, data=None):
        request_kwargs = {
            'url': urljoin(self.hostname, route),
            'headers': {
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {self.secret}',
            }
        }

        if data:
            if method == HttpMethods.GET:
                request_kwargs['params'] = data
            else:
                request_kwargs['json'] = data

        resp = requests.request(method.value, **request_kwargs)

        # raise exception if not success
        try:
            resp.raise_for_status()
        except Exception as exc:
            try:
                raise Exception(resp.json()['message']) from exc
            except ValueError:
                raise Exception(resp.text) from exc

        if resp.status_code == 204:
            return {}

        return resp.json()['data']

File: js/vista-api-client/src/apiResource.js

const HTTP_METHODS = {
    GET: 'GET',
    POST: 'POST',
    DELETE: 'DELETE',
}

class ApiResource {
    constructor(axiosClient, branch, hostname) {
        this.axiosClient = axiosClient;
        this.branch = branch;
        this.hostname = hostname;
    }

    dispatch = async (url, method, data) => {
        const config = {
            url: new URL(url, this.hostname).href,
            method: method,
        }

        if (data) {
            if (method === HTTP_METHODS.GET) {
                config.url = `${config.url}?${new URLSearchParams(data)}`;
            } else if (method === HTTP_METHODS.POST) {
                config.data = data;
            }
        }

        const resp = await this.axiosClient.request(config).catch((error) => {
            if (error.response) {
                throw Error(error.response.data.message);
            } else if (error.request) {
                throw Error('There was a problem with the request');
            } else {
                throw error;
            }
        });

        if (!resp) {
            throw Error('There was a problem with the request');
        }

        if (resp.status === 204) {
            return true;
        }

        return resp.data.data;
    };
}

export { HTTP_METHODS, ApiResource };
